cpp/include/raft/cluster/detail/kmeans_balanced.cuh

```diff
@@ -968,15 +970,15 @@ void build_hierarchical(const raft::resources& handle,
   IdxT n_mesoclusters = std::min(n_clusters, static_cast<IdxT>(std::sqrt(n_clusters) + 0.5));
   RAFT_LOG_DEBUG("build_hierarchical: n_mesoclusters: %u", n_mesoclusters);
 
-  // TODO: Remove the explicit managed memory- we shouldn't be creating this on the user's behalf.
-  rmm::mr::managed_memory_resource managed_memory;
-  rmm::device_async_resource_ref device_memory = resource::get_workspace_resource(handle);
+  rmm::device_async_resource_ref current_device_resource = rmm::mr::get_current_device_resource();
+  rmm::device_async_resource_ref workspace_resource =
```

Comment:
This is one of those rare cases where we do indeed need to explicitly allocate `rmm::mr::managed_memory_resource` (the removed TODO comment is actually incorrect).
The need to use managed memory here has nothing to do with the memory limit and the user choice, but rather is a part of the algorithm. We use the `managed_memory` variable across this file for not-so-big allocations that are accessed by both device and host (see, for example, the `build_fine_clusters` function above). Hence, using the device-only memory simply breaks the algorithm here.

Reply:
Adressed.

However, I'm wondering whether we should be generally using [`make_managed_vector`](https://docs.rapids.ai/api/raft/stable/cpp_api/mdspan_mdarray/#_CPPv4I000EN4raft19make_managed_vectorEDaRKN4raft9resourcesE9IndexType) instead then. @achirkin Was there a specific motivation for the use of `rmm::uvector`s instead?

Reply:
No, only historic reasons: the balanced kmeans code arrived earlier than these managed helpers.

Reply:
Thanks for clarifying.

@cjnolet Considering that this code has moved to cuVS anyways I assume there is no point in refactoring this, is there?

Reply:
No reason to refactor, just need to fix any issues that is blocking cuML UMAP ATM